README


This archive contains the distribution MooseX-Method-Signatures,
version 0.48:

  DEPRECATED: Method declarations with type constraints and no source filter

This software is copyright (c) 2014 by Florian Ragwitz.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for MooseX-Method-Signatures

0.48      2014-09-20 17:46:45Z
  - fix tests that fail on altered warning messages in perl 5.21.4 (see perl
    RT#121638)

0.47      2013-12-14 19:06:59Z
  - update configure_requires checking in Makefile.PL, add CONTRIBUTING file

0.46      2013-10-02 00:40:37Z
  - re-release with fixed compile test

0.45      2013-09-12 03:06:25Z
  - removed use of deprecated Class::MOP::load_class
  - documentation reference added, pointing to the WARNING section in
    MooseX::Declare, regarding future utility of this module
  - repository migrated to the github moose organization

0.44      2012-11-24 21:17:02Z
  - Fix broken test in 5.17.6, where hash ordering is now randomized
    (RT#81388, Karen Etheridge)

0.43      2012-05-21 00:34:04Z
  - Bump dependency on MooseX::Types to 0.35 to resolve issues with recent
    MooseX::Types releases.

0.42      2012-04-23 18:08:19Z
  - Mark failing test due to Eval::Closure/Devel::Declare issues as TODO until
    things are fixed, to at least allow recent versions to install without
    errors.

0.41      2012-03-11 17:05:27Z
  - Add explicit dependency on Eval::Closure, in an effort to get more
    diagnostics from failing cpantesters reports.  No functionality changes.

0.40      2012-03-10 17:03:51Z
  - Loosen a regexp in a test, to handle the new format emitted by Carp 1.25.

0.39      2012-03-10 00:54:18Z
  - The test suite now uses Test::Fatal instead of Test::Exception (Karen
    Etheridge).
  - Bump dependency on MooseX::Types::Structured to 0.24, which contains
    more detailed error messages as well as some performance enhancements.

0.38      2012-03-09 23:43:07Z
  - Require B::Hooks::EndOfScope 0.10 and Module::Runtime 0.012, fixing various
    perl 5.8 breakages.

0.37      2011-09-12 09:35:29Z
  - Don't look in the stash for traits or types that contain :: (Jesse Luehrs).

0.36      2010-08-31 21:36:19Z
  - Avoid deprecation warnings from new Moose version by depending on
    MooseX::LazyRequire 0.06 (Jeremy Carman).

0.35      2010-07-19 03:50:08Z
  - Updated to avoid warnings from Moose 1.09 (Dave Rolsky).
  - Switch from Module::Install to Dist::Zilla.

0.34      2010-06-08 17:00:00Z
  - Provides a version of reify that replaces the original version and
    now finally CatalystX::Declare and MooseX::MultiMethods work.

0.33      2010-06-01 16:47:34Z
  - CatalystX::Declare seems to create a method object without a code
    reference at all.

0.32      2010-06-01 16:47:34Z
  - Detect usage of old MXMS::Meta::Method API and retro-compatibilize.

0.31      2010-06-01 16:47:34Z
  - Add support for "does" as a trait declarator
  - Compatibilize MXMS::Meta::Method signature with Class::MOP::Method
    so other traits can apply customizations to the actual body of the
    method.

0.30      2010-02-04 01:41:36Z
  - Make Any instead of Defined the default type constraint.
  - Add tests for coercing optional named parameters (Cory Watson).
    + Make those tests pass by depending on
      MooseX::Types::Structured 0.20.
  - Make sure Text::Balanced doesn't screw up our data.

0.29      2009-10-15 21:27:08Z
  - Fix regression introduced in 0.28.
    Validation for methods without named arguments wasn't strict enough
    anymore. Passing too many arguments hasn't been rejected. Thanks
    Justin Hunter for finding this.

0.28      2009-10-08 23:28:02Z
  - Fix default values for named arguments.

0.27      2009-09-22 12:30:23Z
  - Depend on MooseX::LazyRequire 0.04 to avoid fail on version 0.03, which has
    been broken by recent Moose versions (Closes RT#49903) (Maik Hentsche).

0.26      2009-09-08 18:14:01Z
  - Rebuilt dist to include missing files

0.25      2009-09-07 21:03:03Z
  - Remove scary ALPHA disclaimer from POD
  - Depend on Test::More 0.88 for done_testing

0.24      2009-08-31 17:04:14Z
  - Rename the meta method's clone method to reify. clone already had a special
    meaning for methods in role applications to classes.

0.23      2009-08-27 20:22:40Z
  - Bump Moose dependency to 0.89. This avoids test failures on some older
    Moose versions.

0.22      2009-08-27 13:29:25Z
  - Provide sane closure behaviour for anonymous methods.
  - Add tests for the above (Sebastian Willert).
  - Depend on MooseX::Types 0.19, as 0.18 made us blow up spectacularly.

0.21      2009-08-18 06:45:48Z
  - Add tests for using aliased.pm for type aliases (Steffen Schwigon).
  - Depend on the latest PMS to get type alias resolving.

0.20      2009-08-17 10:11:32Z
  - Make the meta method's parsed_signature public.
  - Use MooseX::LazyRequire's new interface and depend on 0.03.

0.19      2009-08-14 09:23:42Z
  - Move method redefinition warnings to the method application hook so it's
    easy to override.

0.18      2009-08-14 00:24:41Z
  - Depend on Devel::Declare 0.005011. This avoids "->method" and "method =>"
    being misinterpreted sometimes.

0.17      2009-08-13 06:37:31Z
  - Implement method traits (Cory Watson).
  - Properly test list placeholders.
  - Document the right syntax for return value type constraints.
  - Use namespace::autoclean instead of namespace::clean.
  - Let the signature default to (@) if none is given.
  - Don't require the package attribute when a pre-built context is passed
    in.
  - Depend on a recent Parse::Method::Signatures to make sure we will work
    with new PPIs and have the right MooseX::Traits version.
  - Fix doc typo (arcanez).
  - Split out context from self so that we may pass in an arbitrary context
    (nperez).
  - Enable passing a context and also a coderef to use to apply the method
    once constructed (for deferred method modifier application in MXD)
    (nperez).
  - Enable MXMS to work with multiple different declarators and prototype
    injections (nperez).
  - Change from using Devel::Declare::MethodInstaller::Simple to
    ::Context::Simple since strip_attrs was moved down there (nperez).

0.16      2009-05-01 06:03:44Z

Makefile.PL
# This Makefile.PL for MooseX-Method-Signatures was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.013
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.27.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'ExtUtils::MakeMaker' => '0',
    'Module::Build::Tiny' => '0.038',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.006;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "DEPRECATED: Method declarations with type constraints and no source filter",
  "AUTHOR" => "Florian Ragwitz <rafl\@debian.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build::Tiny" => "0.038"
  },
  "DISTNAME" => "MooseX-Method-Signatures",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MooseX::Method::Signatures",
  "PREREQ_PM" => {
    "B::Hooks::EndOfScope" => "0.10",
    "Carp" => 0,
    "Context::Preserve" => 0,
    "Devel::Declare" => "0.005011",
    "Devel::Declare::Context::Simple" => 0,
    "Eval::Closure" => 0,
    "Module::Runtime" => "0.012",
    "Moose" => "0.89",
    "Moose::Meta::Class" => 0,
    "Moose::Meta::Method" => 0,
    "Moose::Util" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::LazyRequire" => "0.06",
    "MooseX::Meta::TypeConstraint::ForceCoercion" => 0,
    "MooseX::Types" => "0.35",
    "MooseX::Types::Moose" => "0.19",
    "MooseX::Types::Structured" => "0.24",
    "MooseX::Types::Util" => 0,
    "Parse::Method::Signatures" => "1.003014",
    "Parse::Method::Signatures::Param::Named" => 0,
    "Parse::Method::Signatures::Param::Placeholder" => 0,
    "Parse::Method::Signatures::TypeConstraint" => 0,
    "Parse::Method::Signatures::Types" => 0,
    "Scalar::Util" => 0,
    "Sub::Name" => 0,
    "Task::Weaken" => 0,
    "Text::Balanced" => 0,
    "aliased" => 0,
    "if" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Dumper" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Moose::Role" => 0,
    "Test::Deep" => 0,
    "Test::Fatal" => 0,
    "Test::Moose" => 0,
    "Test::More" => "0.89",
    "attributes" => 0,
    "lib" => 0,
    "metaclass" => 0,
    "namespace::clean" => 0
  },
  "VERSION" => "0.48",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "B::Hooks::EndOfScope" => "0.10",
  "Carp" => 0,
  "Context::Preserve" => 0,
  "Data::Dumper" => 0,
  "Devel::Declare" => "0.005011",
  "Devel::Declare::Context::Simple" => 0,
  "Eval::Closure" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Runtime" => "0.012",
  "Moose" => "0.89",
  "Moose::Meta::Class" => 0,
  "Moose::Meta::Method" => 0,
  "Moose::Role" => 0,
  "Moose::Util" => 0,
  "Moose::Util::TypeConstraints" => 0,
  "MooseX::LazyRequire" => "0.06",
  "MooseX::Meta::TypeConstraint::ForceCoercion" => 0,
  "MooseX::Types" => "0.35",
  "MooseX::Types::Moose" => "0.19",
  "MooseX::Types::Structured" => "0.24",
  "MooseX::Types::Util" => 0,
  "Parse::Method::Signatures" => "1.003014",
  "Parse::Method::Signatures::Param::Named" => 0,
  "Parse::Method::Signatures::Param::Placeholder" => 0,
  "Parse::Method::Signatures::TypeConstraint" => 0,
  "Parse::Method::Signatures::Types" => 0,
  "Scalar::Util" => 0,
  "Sub::Name" => 0,
  "Task::Weaken" => 0,
  "Test::Deep" => 0,
  "Test::Fatal" => 0,
  "Test::Moose" => 0,
  "Test::More" => "0.89",
  "Text::Balanced" => 0,
  "aliased" => 0,
  "attributes" => 0,
  "if" => 0,
  "lib" => 0,
  "metaclass" => 0,
  "namespace::autoclean" => 0,
  "namespace::clean" => 0,
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/MooseX/Method/Signatures/Meta/Method.pm
package MooseX::Method::Signatures::Meta::Method;
BEGIN {
  $MooseX::Method::Signatures::Meta::Method::AUTHORITY = 'cpan:ETHER';
}
{
  $MooseX::Method::Signatures::Meta::Method::VERSION = '0.46';
}
# ABSTRACT: Provides the metaclass for methods with signatures

use Moose;
use Carp qw/cluck/;
use Context::Preserve;
use Parse::Method::Signatures 1.003014;
use Parse::Method::Signatures::TypeConstraint;
use Scalar::Util qw/weaken/;
use Moose::Util qw/does_role/;
use Moose::Util::TypeConstraints;
use MooseX::Meta::TypeConstraint::ForceCoercion;
use MooseX::Types::Util qw/has_available_type_export/;
use MooseX::Types::Structured qw/Tuple Dict Optional slurpy/;
use MooseX::Types::Moose qw/ArrayRef Str Maybe Object Any CodeRef Bool/;
use MooseX::Method::Signatures::Types qw/Injections Params/;
use aliased 'Parse::Method::Signatures::Param::Named';
use aliased 'Parse::Method::Signatures::Param::Placeholder';

use namespace::autoclean;

extends 'Moose::Meta::Method';

has signature => (
    is       => 'ro',
    isa      => Str,
    default  => '(@)',
    required => 1,
);

has parsed_signature => (
    is      => 'ro',
    isa     => class_type('Parse::Method::Signatures::Sig'),
    lazy    => 1,
    builder => '_build_parsed_signature',
);

sub _parsed_signature {
    cluck '->_parsed_signature is deprecated. use ->parsed_signature instead.';
    shift->parsed_signature;
}

has _lexicals => (
    is      => 'ro',
    isa     => ArrayRef[Str],
    lazy    => 1,
    builder => '_build__lexicals',
);

has injectable_code => (
    is      => 'ro',
    isa     => Str,
    lazy    => 1,
    builder => '_build_injectable_code',
);

has _positional_args => (
    is      => 'ro',
    isa     => ArrayRef,
    lazy    => 1,
    builder => '_build__positional_args',
);

has _named_args => (
    is      => 'ro',
    isa     => ArrayRef,
    lazy    => 1,
    builder => '_build__named_args',
);

has _has_slurpy_positional => (
    is   => 'rw',
    isa  => Bool,
);

has type_constraint => (
    is      => 'ro',
    isa     => class_type('Moose::Meta::TypeConstraint'),
    lazy    => 1,
    builder => '_build_type_constraint',
);

has return_signature => (
    is        => 'ro',
    isa       => Str,
    predicate => 'has_return_signature',
);

has _return_type_constraint => (
    is      => 'ro',
    isa     => class_type('Moose::Meta::TypeConstraint'),
    lazy    => 1,
    builder => '_build__return_type_constraint',
);

has actual_body => (
    is        => 'ro',
    isa       => CodeRef,
    predicate => '_has_actual_body',
);

has prototype_injections => (
    is          => 'rw',
    isa         => Injections,
    trigger     => \&_parse_prototype_injections
);

has _parsed_prototype_injections => (
    is          => 'ro',
    isa         => Params,
    predicate   => '_has_parsed_prototype_injections',
    writer      => '_set_parsed_prototype_injections',
);

before actual_body => sub {
    my ($self) = @_;
    confess "method doesn't have an actual body yet"
        unless $self->_has_actual_body;
};

around name => sub {
    my ($next, $self) = @_;
    my $ret = $self->$next;
    confess "method doesn't have a name yet"
        unless defined $ret;
    return $ret;
};

sub _wrapped_body {
    my ($class, $self, %args) = @_;

    if (exists $args{return_signature}) {
        return sub {
            my @args = ${ $self }->validate(\@_);
            return preserve_context { ${ $self }->actual_body->(@args) }
                after => sub {
                    if (defined (my $msg = ${ $self }->_return_type_constraint->validate(\@_))) {
                        confess $msg;
                    }
                };
        };
    }

    my $actual_body;
    return sub {
        @_ = ${ $self }->validate(\@_);
        $actual_body ||= ${ $self }->actual_body;
        goto &{ $actual_body };
    };

}

around wrap => sub {
    my $orig = shift;
    my $self;
    my ($class, $code, %args);
    if (ref $_[1]) {
        ($class, $code, %args) = @_;
    } else {
        ($class, %args) = @_;
        $code = delete $args{body};
    }

    my $wrapped = $class->_wrapped_body(\$self, %args);
    $self = $class->$orig($wrapped, %args, $code ? (actual_body => $code) : ());

    # Vivify the type constraints so TC lookups happen before namespace::clean
    # removes them
    $self->type_constraint;
    $self->_return_type_constraint if $self->has_return_signature;

    weaken($self->{associated_metaclass})
        if $self->{associated_metaclass};

    return $self;
};

sub reify {
    my $self = shift;
    my %args = @_;

    my %other_args = %{$self};
    delete $other_args{body};
    delete $other_args{actual_body};

    my $body = delete $args{body} || delete $args{actual_body} || $self->body;
    my %final_args = (%other_args, %args);

    return $self->meta->name->wrap($body, %final_args);
}

sub _build_parsed_signature {
    my ($self) = @_;
    return Parse::Method::Signatures->signature(
        input => $self->signature,
        from_namespace => $self->package_name,
    );
}

sub _build__return_type_constraint {
    my ($self) = @_;
    confess 'no return type constraint'
        unless $self->has_return_signature;

    my $parser = Parse::Method::Signatures->new(
        input => $self->return_signature,
        from_namespace => $self->package_name,
    );

    my $param = $parser->_param_typed({});
    confess 'failed to parse return value type constraint'
        unless exists $param->{type_constraints};

    return Tuple[$param->{type_constraints}->tc];
}

sub _param_to_spec {
    my ($self, $param) = @_;

    my $tc = Any;
    {
        # Ensure errors get reported from the right place
        local $Carp::Internal{'MooseX::Method::Signatures::Meta::Method'} = 1;
        local $Carp::Internal{'Moose::Meta::Method::Delegation'} = 1;
        local $Carp::Internal{'Moose::Meta::Method::Accessor'} = 1;
        local $Carp::Internal{'MooseX::Method::Signatures'} = 1;
        local $Carp::Internal{'Devel::Declare'} = 1;
        $tc = $param->meta_type_constraint
          if $param->has_type_constraints;
    }

    if ($param->has_constraints) {
        my $cb = join ' && ', map { "sub {${_}}->(\\\@_)" } $param->constraints;
        my $code = eval "sub {${cb}}";
        $tc = subtype({ as => $tc, where => $code });
    }

    my %spec;
    if ($param->sigil ne '$') {
        $spec{slurpy} = 1;
        $tc = slurpy ArrayRef[$tc];
    }

    $spec{tc} = $param->required
        ? $tc
        : Optional[$tc];

    $spec{default} = $param->default_value
        if $param->has_default_value;

    if ($param->has_traits) {
        for my $trait (@{ $param->param_traits }) {
            next unless $trait->[1] eq 'coerce';
            $spec{coerce} = 1;
        }
    }

    return \%spec;
}

sub _parse_prototype_injections {
    my $self = shift;

    my @params;
    for my $inject (@{ $self->prototype_injections }) {
        my $param;
        eval {
            $param = Parse::Method::Signatures->param($inject);
        };

        confess "There was a problem parsing the prototype injection '$inject': $@"
            if $@ || !defined $param;

        push @params, $param;
    }

    my @return = reverse @params;
    $self->_set_parsed_prototype_injections(\@return);
}

sub _build__lexicals {
    my ($self) = @_;
    my ($sig) = $self->parsed_signature;

    my @lexicals;

    if ($self->_has_parsed_prototype_injections) {
        push @lexicals, $_->variable_name
            for @{ $self->_parsed_prototype_injections };
    }

    push @lexicals, $sig->has_invocant
        ? $sig->invocant->variable_name
        : '$self';

    push @lexicals,
        (does_role($_, Placeholder)
            ? 'undef'
            : $_->variable_name)
        for (($sig->has_positional_params ? $sig->positional_params : ()),
             ($sig->has_named_params      ? $sig->named_params      : ()));

    return \@lexicals;
}

sub _build_injectable_code {
    my ($self) = @_;
    my $vars = join q{,}, @{ $self->_lexicals };
    return "my (${vars}) = \@_;";
}

sub _build__positional_args {
    my ($self) = @_;
    my $sig = $self->parsed_signature;

    my @positional;
    if ($self->_has_parsed_prototype_injections) {
        push @positional, map {
            $self->_param_to_spec($_)
        } @{ $self->_parsed_prototype_injections };
    }

    push @positional, $sig->has_invocant
        ? $self->_param_to_spec($sig->invocant)
        : { tc => Object };

    my $slurpy = 0;
    if ($sig->has_positional_params) {
        for my $param ($sig->positional_params) {
            my $spec = $self->_param_to_spec($param);
            $slurpy ||= 1 if $spec->{slurpy};
            push @positional, $spec;
        }
    }

    $self->_has_slurpy_positional($slurpy);
    return \@positional;
}

sub _build__named_args {
    my ($self) = @_;
    my $sig = $self->parsed_signature;

    # triggering building of positionals before named params is important
    # because the latter needs to know if there have been any slurpy
    # positionals to report errors
    $self->_positional_args;

    my @named;

    if ($sig->has_named_params) {
        confess 'Named parameters cannot be combined with slurpy positionals'
            if $self->_has_slurpy_positional;
        for my $param ($sig->named_params) {
            push @named, $param->label => $self->_param_to_spec($param);
        }
    }

    return \@named;
}

sub _build_type_constraint {
    my ($self) = @_;
    my ($positional, $named) = map { $self->$_ } map { "_${_}_args" } qw/positional named/;

    my $tc = Tuple[
        Tuple[ map { $_->{tc}               } @{ $positional } ],
        Dict[  map { ref $_ ? $_->{tc} : $_ } @{ $named      } ],
    ];

    my $coerce_param = sub {
        my ($spec, $value) = @_;
        return $value unless exists $spec->{coerce};
        return $spec->{tc}->coerce($value);
    };

    my %named = @{ $named };

    coerce $tc,
        from ArrayRef,
        via {
            my (@positional_args, %named_args);

            my $i = 0;
            for my $param (@{ $positional }) {
                push @positional_args, map { $coerce_param->($param, $_) }
                    $#{ $_ } < $i
                        ? (exists $param->{default} ? eval $param->{default} : ())
                        : $_->[$i];
                $i++;
            }

            if (%named) {
                my @rest = @{ $_ }[$i .. $#{ $_ }];
                confess "Expected named arguments but didn't find an even-sized list"
                    unless @rest % 2 == 0;
                my %rest = @rest;

                while (my ($key, $spec) = each %named) {
                    if (exists $rest{$key}) {
                        $named_args{$key} = $coerce_param->($spec, delete $rest{$key});
                        next;
                    }

                    if (exists $spec->{default}) {
                        $named_args{$key} = $coerce_param->($spec, eval $spec->{default});
                    }
                }

                @named_args{keys %rest} = values %rest;
            }
            elsif ($#{ $_ } >= $i) {
                push @positional_args, @{ $_ }[$i .. $#{ $_ }];
            }

            return [\@positional_args, \%named_args];
        };

    return MooseX::Meta::TypeConstraint::ForceCoercion->new(
        type_constraint => $tc,
    );
}

sub validate {
    my ($self, $args) = @_;

    my @named = grep { !ref $_ } @{ $self->_named_args };

    my $coerced;
    if (defined (my $msg = $self->type_constraint->validate($args, \$coerced))) {
        confess $msg;
    }

    return @{ $coerced->[0] }, map { $coerced->[1]->{$_} } @named;
}

__PACKAGE__->meta->make_immutable;


1;

__END__

=pod

=encoding ISO-8859-1

=for :stopwords Florian Ragwitz metaclass

=head1 NAME

MooseX::Method::Signatures::Meta::Method - Provides the metaclass for methods with signatures

=head1 VERSION

version 0.46

=head1 AUTHOR

Florian Ragwitz <rafl@debian.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Florian Ragwitz.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL
# This Build.PL for MooseX-Method-Signatures was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.007
use strict;
use warnings;

if (eval 'use Module::Build::Tiny 0.038; 1')
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";

    # This section for MooseX-Method-Signatures was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.007.
    use strict;
    use warnings;

    use 5.006;
    # use Module::Build::Tiny 0.038;
    Build_PL();
}
else
{
    $ENV{PERL_MB_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.

EOW
    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "build_requires" => {},
      "configure_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "Module::Build::Tiny" => "0.038"
      },
      "dist_abstract" => "DEPRECATED: Method declarations with type constraints and no source filter",
      "dist_author" => [
        "Florian Ragwitz <rafl\@debian.org>"
      ],
      "dist_name" => "MooseX-Method-Signatures",
      "dist_version" => "0.48",
      "license" => "perl",
      "module_name" => "MooseX::Method::Signatures",
      "recommends" => {},
      "recursive_test_files" => 1,
      "requires" => {
        "B::Hooks::EndOfScope" => "0.10",
        "Carp" => 0,
        "Context::Preserve" => 0,
        "Devel::Declare" => "0.005011",
        "Devel::Declare::Context::Simple" => 0,
        "Eval::Closure" => 0,
        "Module::Runtime" => "0.012",
        "Moose" => "0.89",
        "Moose::Meta::Class" => 0,
        "Moose::Meta::Method" => 0,
        "Moose::Util" => 0,
        "Moose::Util::TypeConstraints" => 0,
        "MooseX::LazyRequire" => "0.06",
        "MooseX::Meta::TypeConstraint::ForceCoercion" => 0,
        "MooseX::Types" => "0.35",
        "MooseX::Types::Moose" => "0.19",
        "MooseX::Types::Structured" => "0.24",
        "MooseX::Types::Util" => 0,
        "Parse::Method::Signatures" => "1.003014",
        "Parse::Method::Signatures::Param::Named" => 0,
        "Parse::Method::Signatures::Param::Placeholder" => 0,
        "Parse::Method::Signatures::TypeConstraint" => 0,
        "Parse::Method::Signatures::Types" => 0,
        "Scalar::Util" => 0,
        "Sub::Name" => 0,
        "Task::Weaken" => 0,
        "Text::Balanced" => 0,
        "aliased" => 0,
        "if" => 0,
        "namespace::autoclean" => 0,
        "perl" => "5.006",
        "strict" => 0,
        "warnings" => 0
      },
      "script_files" => [],
      "test_requires" => {
        "Data::Dumper" => 0,
        "ExtUtils::MakeMaker" => 0,
        "File::Spec" => 0,
        "Moose::Role" => 0,
        "Test::Deep" => 0,
        "Test::Fatal" => 0,
        "Test::Moose" => 0,
        "Test::More" => "0.89",
        "attributes" => 0,
        "lib" => 0,
        "metaclass" => 0,
        "namespace::clean" => 0
      }
    );


    my %fallback_build_requires = (
      "Data::Dumper" => 0,
      "ExtUtils::MakeMaker" => 0,
      "File::Spec" => 0,
      "Moose::Role" => 0,
      "Test::Deep" => 0,
      "Test::Fatal" => 0,
      "Test::Moose" => 0,
      "Test::More" => "0.89",
      "attributes" => 0,
      "lib" => 0,
      "metaclass" => 0,
      "namespace::clean" => 0
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);

    $build->create_build_script;
}
